### PR TITLE
Consumer load metrics for weighted work balancer

### PR DESCRIPTION
**Pull request by piotrrzysko**
This PR introduces consumer load metrics that will be used by the weighted work balancer. For now, I want to start with one metric: the number of operations per second that a given consumer performs. An operation is a unit of work that a consumer executes in the context of a subscription, such as receiving a message from Kafka, sending a message to the subscriber, or handling a signal.

### CODE REVIEW

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/weighted/ConsumerNodeLoadDecoder.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    private static final Logger logger = getLogger(ConsumerNodeLoadDecoder.class);
+
+    private final SubscriptionIds subscriptionIds;
+
+    ConsumerNodeLoadDecoder(SubscriptionIds subscriptionIds) {
+        this.subscriptionIds = subscriptionIds;
+    }
+
+    ConsumerNodeLoad decode(byte[] bytes) {
```
Comment (szczygiel-m) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/weighted/ConsumerNodeLoadDecoder.java:28]: What do You think about splitting this function?

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/weighted/ConsumerNodeLoadEncoder.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import pl.allegro.tech.hermes.consumers.supervisor.workload.sbe.stubs.ConsumerLoadEncoder.SubscriptionsEncoder;
+import pl.allegro.tech.hermes.consumers.supervisor.workload.sbe.stubs.MessageHeaderEncoder;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+class ConsumerNodeLoadEncoder {
+
+    private static final int BUFFER_SIZE_IN_BYTES = 100_000;
```
Comment (szczygiel-m) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/weighted/ConsumerNodeLoadEncoder.java:18]: Don't we want to add this as a configuration parameter?

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/weighted/NoOpConsumerNodeLoadRegistry.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    public ConsumerNodeLoad get(String consumerId) {
+        return ConsumerNodeLoad.UNDEFINED;
+    }
+
+    @Override
+    public SubscriptionLoadRecorder register(SubscriptionName subscriptionName) {
+        return SUBSCRIPTION_LOAD_RECORDER;
+    }
+
+    private static class NoOpSubscriptionLoadRecorder implements SubscriptionLoadRecorder {
```
Comment (szczygiel-m) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/weighted/NoOpConsumerNodeLoadRegistry.java:30]: This implementation is repeated here and in test directory, what do You think about removing one?

**Review by szczygiel-m — COMMENTED** [review]
Nice PR 💪 added some questions and minor comments

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java [hunk 4]
```diff
@@ -123,10 +127,25 @@ private void offerCommittedOffsets(MessageBatch batch) {
...
+                receiver,
+                batchFactory,
+                hermesMetrics,
+                messageConverterResolver,
+                compositeMessageContentWrapper,
+                topic,
+                trackers,
+                loadRecorder
+        );
+        loadRecorder.initialize();
```
[the remark this answers is not in the document]
Comment (piotrrzysko) [reply hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: Good question. In terms of correctness there is no big difference, but more logically is to call this method before passing the object to the receivers. Fixed.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/weighted/ConsumerNodeLoadEncoder.java (hunk 2 above)
Comment (piotrrzysko) [reply hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/weighted/ConsumerNodeLoadEncoder.java]: I was thinking about it. I know that we do so in other encoders. The only problem I have with this parameter is that it is another one in the long list of parameters we already have. Worse yet, it is unclear how to properly set it up. Maybe in the future we can get rid of them and automatically determine their values. 

Anyway, I think your suggestion is reasonable, so I made it configurable.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/weighted/NoOpConsumerNodeLoadRegistry.java (hunk 3 above)
Comment (piotrrzysko) [reply hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/weighted/NoOpConsumerNodeLoadRegistry.java:30]: Good point. Fixed.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/weighted/ConsumerNodeLoadDecoder.java (hunk 1 above)
Comment (piotrrzysko) [reply hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/weighted/ConsumerNodeLoadDecoder.java:28]: Done.

**Review by szczygiel-m — APPROVED** [review]

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/time/ModifiableClock.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
+package pl.allegro.tech.hermes.common.time;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.temporal.ChronoUnit;
+
+import static java.time.Instant.now;
+
+public class ModifiableClock extends Clock {
```
Comment (faderskd) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/time/ModifiableClock.java:10]: I guess this class should be in tests package

Comment (piotrrzysko) [reply hermes-common/src/main/java/pl/allegro/tech/hermes/common/time/ModifiableClock.java]: Indeed. Fixed. Thanks for spotting that.

**Review by szczygiel-m — APPROVED** [review]

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/weighted/ZookeeperConsumerNodeLoadRegistry.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+            persist(consumerNodeLoad);
+        } catch (Exception e) {
+            logger.error("Error while reporting consumer node load", e);
+        }
+    }
+
+    private ConsumerNodeLoad calculateConsumerNodeLoad() {
+        long now = clock.millis();
+        long elapsedMillis = now - lastReset;
+        long elapsedSeconds = Math.max(MILLISECONDS.toSeconds(elapsedMillis), 1);
```
Comment (faderskd) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/weighted/ZookeeperConsumerNodeLoadRegistry.java:96]: shouldn't it be `Math.min(...)`  ?

**Review by faderskd — APPROVED** [review]

**Pull request merged**